File: obs_replaykit/gpu.py
"""detect the users gpu(s) and classify them for encoder selection. pulls win32_videocontroller via cim/wmi and parses the adapter name to tag the architecture tier (registry/wmi doesnt expose the arch as a structured field, but marketing names follow stable patterns). prefers discrete over integrated when both are present."""

from __future__ import annotations

import re
import subprocess
from dataclasses import dataclass
from enum import Enum
from typing import List, Optional


class Vendor(str, Enum):
    NVIDIA  = "nvidia"
    AMD     = "amd"
    INTEL   = "intel"
    UNKNOWN = "unknown"


class NvidiaGen(str, Enum):
    """coarse nvenc capability tiers."""
    PRE_MAXWELL2 = "pre_maxwell2"  # kepler and older -- no hevc encode
    MAXWELL2     = "maxwell2"  # gtx 9xx -- hevc, no b-frames, qres only
    PASCAL       = "pascal"  # gtx 10xx -- hevc + lookahead, no hevc b-frames
    TURING       = "turing"  # gtx 16xx / rtx 20xx -- hevc b-frames
    AMPERE       = "ampere"  # rtx 30xx
    ADA          = "ada"  # rtx 40xx
    BLACKWELL    = "blackwell"  # rtx 50xx -- av1 nvenc, nvenc 10th gen


class AmdGen(str, Enum):
    """coarse amf capability tiers."""
    PRE_POLARIS = "pre_polaris"  # no hardware hevc
    POLARIS     = "polaris"  # rx 4/5xx -- first amd hevc (basic)
    VEGA        = "vega"  # vega / radeon vii
    RDNA1       = "rdna1"  # rx 5xxx
    RDNA2       = "rdna2"  # rx 6xxx
    RDNA3       = "rdna3"  # rx 7xxx -- av1 amf
    RDNA4       = "rdna4"  # rx 9xxx (future-proof)


class IntelGen(str, Enum):
    """coarse quick sync capability tiers."""
    PRE_SKYLAKE = "pre_skylake"  # no qsv hevc
    SKYLAKE     = "skylake"  # hd 530 -- first qsv hevc, 8-bit
    KABY_LAKE   = "kaby_lake"  # hd 630 -- + 10-bit hevc
    ICE_LAKE    = "ice_lake"  # big qsv quality jump
    TIGER_LAKE  = "tiger_lake"  # xe-lp igpu
    ALDER_LAKE  = "alder_lake"  # 12th/13th/14th gen igpu
    ARC         = "arc"  # discrete arc + xe-hpg (av1 qsv)


@dataclass(frozen=True)
class Gpu:
    """one physical adapter detected on the system."""
    name:         str  # win32_videocontroller.name
    vendor:       Vendor
    generation:   Optional[str]  # nvidiagen / amdgen / intelgen value (string)
    is_discrete:  bool  # heuristic: discrete dedicated card vs igpu
    vram_bytes:   int  # adapterram (0 if unknown)
    driver:       str  # driverversion

    @property
    def label(self) -> str:
        return self.name

    @property
    def gen_label(self) -> str:
        if self.generation is None:
            return "unknown"
        return self.generation.replace("_", " ").title()


def _query_video_controllers() -> List[dict]:
    """win32_videocontroller rows as plain dicts. powershell get-ciminstance becuase wmic was deprecated in win10 21h1 and isnt installed on win11. returns [] on any failure."""
    cmd = [
        "powershell.exe",
        "-NoProfile",
        "-NonInteractive",
        "-Command",
        # @() forces array semantics even with one controller (otherwise convertto-json serialises a single object as a bare dict). -asarray would also work but its powershell 7+; we still run on stock powershell 5.1.
        "ConvertTo-Json -Depth 3 -Compress -InputObject @("
        "Get-CimInstance Win32_VideoController | "
        "Select-Object Name, AdapterRAM, DriverVersion)",
    ]
    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True,
            timeout=10, creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0),
        )
    except (OSError, subprocess.TimeoutExpired):
        return []
    if result.returncode != 0 or not result.stdout.strip():
        return []
    try:
        import json
        data = json.loads(result.stdout)
    except (json.JSONDecodeError, ValueError):
        return []
    if not isinstance(data, list):
        return [data] if isinstance(data, dict) else []
    return data


_VENDOR_PATTERNS = (
    (re.compile(r"\bnvidia\b|\bgeforce\b|\bquadro\b|\brtx\b|\bgtx\b", re.IGNORECASE), Vendor.NVIDIA),
    (re.compile(r"\bamd\b|\bradeon\b|\brx\s*\d", re.IGNORECASE),                      Vendor.AMD),
    (re.compile(r"\bintel\b|\biris\b|\barc\b|\buhd\b|\bhd\s*graphics\b", re.IGNORECASE), Vendor.INTEL),
)

# basic-display-adapter / indirect-display / remote-display drivers we want to ignore (no real encoder support, never the right answer for obs).
_IGNORE_NAME_PATTERNS = (
    re.compile(r"\bbasic\s+display\s+adapter\b", re.IGNORECASE),
    re.compile(r"\bidd\b|indirect\s+display", re.IGNORECASE),
    re.compile(r"\bremote\s+display\b", re.IGNORECASE),
    re.compile(r"\bvirtual\s+display\b", re.IGNORECASE),
    re.compile(r"\bdisplay\s*link\b", re.IGNORECASE),
)


def _detect_vendor(name: str) -> Vendor:
    for pattern, vendor in _VENDOR_PATTERNS:
        if pattern.search(name):
            return vendor
    return Vendor.UNKNOWN


# nvidia model -> arch. ordered most-specific first; first match wins.
_NVIDIA_PATTERNS = (
    # blackwell -- rtx 50-series (2025+). 50xx covers desktop + mobile.
    (re.compile(r"\brtx\s*50\d{2}\b", re.IGNORECASE),                                   NvidiaGen.BLACKWELL),
    # ada lovelace -- rtx 40-series (2022-2024).
    (re.compile(r"\brtx\s*40\d{2}\b", re.IGNORECASE),                                   NvidiaGen.ADA),
    # ampere -- rtx 30-series.
    (re.compile(r"\brtx\s*30\d{2}\b", re.IGNORECASE),                                   NvidiaGen.AMPERE),
    # turing -- rtx 20-series + gtx 16-series.
    (re.compile(r"\brtx\s*20\d{2}\b|\bgtx\s*16\d{2}\b|\btitan\s*rtx\b", re.IGNORECASE), NvidiaGen.TURING),
    # pascal -- gtx 10-series + titan xp; pascal quadro variants match via the 'p' prefix.
    (re.compile(r"\bgtx\s*10\d{2}\b|\btitan\s*xp\b|\bquadro\s*p\d", re.IGNORECASE),     NvidiaGen.PASCAL),
    # maxwell 2nd gen -- gtx 9-series (gm204/206). gtx 750/750ti was maxwell 1 which didnt have hevc, so we deliberately dont claim it.
    (re.compile(r"\bgtx\s*9\d{2}\b|\btitan\s*x\b", re.IGNORECASE),                      NvidiaGen.MAXWELL2),
)


def _detect_nvidia_gen(name: str) -> NvidiaGen:
    for pattern, gen in _NVIDIA_PATTERNS:
        if pattern.search(name):
            return gen
    return NvidiaGen.PRE_MAXWELL2


_AMD_PATTERNS = (
    # rdna 4 -- rx 9000-series (forward-looking).
    (re.compile(r"\brx\s*9\d{3}\b", re.IGNORECASE),                                                  AmdGen.RDNA4),
    # rdna 3 -- rx 7000-series.
    (re.compile(r"\brx\s*7\d{3}\b", re.IGNORECASE),                                                  AmdGen.RDNA3),
    # rdna 2 -- rx 6000-series + rdna-2 integrated (rembrandt, ryzen 6000u/h).
    (re.compile(r"\brx\s*6\d{3}\b|\bradeon\s+graphics\s*\(rembrandt\)", re.IGNORECASE),              AmdGen.RDNA2),
    # rdna 1 -- rx 5000-series.
    (re.compile(r"\brx\s*5\d{3}\b", re.IGNORECASE),                                                  AmdGen.RDNA1),
    # vega / radeon vii.
    (re.compile(r"\brx\s+vega\b|\bvega\s*\d|\bradeon\s+vii\b", re.IGNORECASE),                       AmdGen.VEGA),
    # polaris -- rx 4xx / 5xx (570/580/590 etc.). 5xx range conflicts with rdna1s 5000-series, so we match 3-digit '5xx' only (not 4-digit '5xxx').
    (re.compile(r"\brx\s*4\d{2}\b|\brx\s*5\d{2}(?!\d)", re.IGNORECASE),                              AmdGen.POLARIS),
)


def _detect_amd_gen(name: str) -> AmdGen:
    for pattern, gen in _AMD_PATTERNS:
        if pattern.search(name):
            return gen
    return AmdGen.PRE_POLARIS


_INTEL_PATTERNS = (
    # arc discrete + xe-hpg. arc a310/a380/a580/a750/a770 all share the alchemist hevc encode block.
    (re.compile(r"\barc\s*a?\d{3}\b|\bxe-hpg\b", re.IGNORECASE),                IntelGen.ARC),
    # alder/raptor/meteor lake iris xe in 12th-14th gen.
    (re.compile(r"\b(12|13|14)\d{2}[a-z]?\b.*(uhd|iris)", re.IGNORECASE),       IntelGen.ALDER_LAKE),
    # tiger lake (11th gen) -- xe-lp.
    (re.compile(r"\b11\d{2}[a-z]?\b.*(iris|xe)|\bxe\s*graphics\b", re.IGNORECASE), IntelGen.TIGER_LAKE),
    # ice lake (10th gen) -- first xe igpu.
    (re.compile(r"\b10\d{2}[a-z]?\b.*iris", re.IGNORECASE),                     IntelGen.ICE_LAKE),
    # kaby lake (7th gen) hd 6xx. coffee lake (8th/9th) + comet lake (10th non-ice-lake) reuse the same igpu silicon, so they share this rule.
    (re.compile(r"\b(uhd|hd)\s*graphics\s*6\d{2}\b", re.IGNORECASE),            IntelGen.KABY_LAKE),
    # skylake (6th gen) hd 5xx -- first with qsv hevc.
    (re.compile(r"\bhd\s*graphics\s*5\d{2}\b", re.IGNORECASE),                  IntelGen.SKYLAKE),
)


def _detect_intel_gen(name: str) -> IntelGen:
    for pattern, gen in _INTEL_PATTERNS:
        if pattern.search(name):
            return gen
    return IntelGen.PRE_SKYLAKE


def _classify(raw: dict) -> Optional[Gpu]:
    """map a win32_videocontroller row to a Gpu, or None if it looks like a virtual/basic/fallback adapter we should ignore."""
    name = (raw.get("Name") or "").strip()
    if not name:
        return None
    for skip in _IGNORE_NAME_PATTERNS:
        if skip.search(name):
            return None

    vendor = _detect_vendor(name)
    if vendor is Vendor.NVIDIA:
        gen = _detect_nvidia_gen(name).value
    elif vendor is Vendor.AMD:
        gen = _detect_amd_gen(name).value
    elif vendor is Vendor.INTEL:
        gen = _detect_intel_gen(name).value
    else:
        gen = None

    vram = int(raw.get("AdapterRAM") or 0)
    # heuristic: igpus typically expose <2 gb adapterram (rest is shared/dynamic). discrete cards report full vram. 2 gb threshold catches apus/igpus that reserve up to 2 gb dedicated (ryzen) without losing genuinely small discrete cards from ~2014.
    is_discrete = vram >= 2 * 1024**3 and vendor is not Vendor.INTEL

    return Gpu(
        name        = name,
        vendor      = vendor,
        generation  = gen,
        is_discrete = is_discrete,
        vram_bytes  = vram,
        driver      = (raw.get("DriverVersion") or "").strip(),
    )


def list_gpus() -> List[Gpu]:
    """every adapter we can usefully classify, discrete cards first."""
    rows = _query_video_controllers()
    out  = [g for g in (_classify(r) for r in rows) if g is not None]
    # sort: discrete first; within tier, nvidia > amd > intel > unknown.
    vendor_rank = {Vendor.NVIDIA: 0, Vendor.AMD: 1, Vendor.INTEL: 2, Vendor.UNKNOWN: 3}
    out.sort(key=lambda g: (not g.is_discrete, vendor_rank[g.vendor]))
    return out


def primary_gpu() -> Optional[Gpu]:
    """best guess at which adapter obs will use for hardware encoding. None means callers should fall back to software encode."""
    gpus = list_gpus()
    return gpus[0] if gpus else None

File: ReplayKit/ReplayKitHelper/Gpu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Management;
using System.Text.RegularExpressions;
using Newtonsoft.Json.Linq;

namespace ReplayKitHelper
{
    // detect the users gpu(s) and classify them for encoder selection. pulls win32_videocontroller via wmi and parses the adapter name to tag the architecture tier (registry/wmi doesnt expose the arch as a structured field, but marketing names follow stable patterns). prefers discrete over integrated when both are present. this is the exact same detection logic as ReplayKitSetup/Gpu.cs (the fresh-install path) -- copied rather than re-derived from obs_replaykit helper modules/62_replaykit_settings.ps1s Get-ReplayKitPrimaryGpu so both the installer and the live settings dock agree on the same classification, and so a bug fixed in one doesnt need re-fixing in the other by hand. only real change: QueryVideoControllers queries WMI directly instead of spawning powershell.exe, since this project already references System.Management (ReplayKitSetup didnt, historically).
    public enum Vendor { Nvidia, Amd, Intel, Unknown }

    // coarse nvenc capability tiers.
    public enum NvidiaGen { PreMaxwell2, Maxwell2, Pascal, Turing, Ampere, Ada, Blackwell }

    // coarse amf capability tiers.
    public enum AmdGen { PrePolaris, Polaris, Vega, Rdna1, Rdna2, Rdna3, Rdna4 }

    // coarse quick sync capability tiers.
    public enum IntelGen { PreSkylake, Skylake, KabyLake, IceLake, TigerLake, AlderLake, Arc }

    // one physical adapter detected on the system.
    public sealed class GpuInfo
    {
        public string Name { get; set; }
        public Vendor Vendor { get; set; }
        public string Generation { get; set; } // nvidiagen/amdgen/intelgen snake_case value, or null
        public bool IsDiscrete { get; set; }
        public long VramBytes { get; set; }
        public string Driver { get; set; }

        public string Label => Name;

        public string GenLabel
        {
            get
            {
                if (Generation == null) return "unknown";
                var words = Generation.Split('_').Select(w => w.Length == 0 ? w : char.ToUpperInvariant(w[0]) + w.Substring(1));
                return string.Join(" ", words);
            }
        }
    }

    public static class Gpu
    {
        private static List<JObject> QueryVideoControllers()
        {
            var result = new List<JObject>();
            try
            {
                using (var searcher = new ManagementObjectSearcher("SELECT Name, AdapterRAM, DriverVersion FROM Win32_VideoController"))
                using (var rows = searcher.Get())
                {
                    foreach (ManagementObject row in rows)
                    {
                        long? ram = null;
                        try { if (row["AdapterRAM"] != null) ram = Convert.ToInt64(row["AdapterRAM"]); } catch (Exception ex) when (ex is InvalidCastException || ex is OverflowException) { }
                        result.Add(new JObject
                        {
                            ["Name"] = row["Name"] as string,
                            ["AdapterRAM"] = ram,
                            ["DriverVersion"] = row["DriverVersion"] as string,
                        });
                    }
                }
            }
            catch (ManagementException) { }
            return result;
        }

        private static readonly (Regex Pattern, Vendor Vendor)[] VendorPatterns =
        {
            (new Regex(@"\bnvidia\b|\bgeforce\b|\bquadro\b|\brtx\b|\bgtx\b", RegexOptions.IgnoreCase), Vendor.Nvidia),
            (new Regex(@"\bamd\b|\bradeon\b|\brx\s*\d", RegexOptions.IgnoreCase), Vendor.Amd),
            (new Regex(@"\bintel\b|\biris\b|\barc\b|\buhd\b|\bhd\s*graphics\b", RegexOptions.IgnoreCase), Vendor.Intel),
        };

        // basic-display-adapter / indirect-display / remote-display drivers we want to ignore (no real encoder support, never the right answer for obs).
        private static readonly Regex[] IgnoreNamePatterns =
        {
            new Regex(@"\bbasic\s+display\s+adapter\b", RegexOptions.IgnoreCase),
            new Regex(@"\bidd\b|indirect\s+display", RegexOptions.IgnoreCase),
            new Regex(@"\bremote\s+display\b", RegexOptions.IgnoreCase),
            new Regex(@"\bvirtual\s+display\b", RegexOptions.IgnoreCase),
            new Regex(@"\bdisplay\s*link\b", RegexOptions.IgnoreCase),
        };

        private static Vendor DetectVendor(string name)
        {
            foreach (var (pattern, vendor) in VendorPatterns)
            {
                if (pattern.IsMatch(name)) return vendor;
            }
            return Vendor.Unknown;
        }

        // nvidia model -> arch. ordered most-specific first; first match wins.
        private static readonly (Regex Pattern, NvidiaGen Gen)[] NvidiaPatterns =
        {
            // blackwell -- rtx 50-series (2025+). 50xx covers desktop + mobile.
            (new Regex(@"\brtx\s*50\d{2}\b", RegexOptions.IgnoreCase), NvidiaGen.Blackwell),
            // ada lovelace -- rtx 40-series (2022-2024).
            (new Regex(@"\brtx\s*40\d{2}\b", RegexOptions.IgnoreCase), NvidiaGen.Ada),
            // ampere -- rtx 30-series.
            (new Regex(@"\brtx\s*30\d{2}\b", RegexOptions.IgnoreCase), NvidiaGen.Ampere),
            // turing -- rtx 20-series + gtx 16-series.
            (new Regex(@"\brtx\s*20\d{2}\b|\bgtx\s*16\d{2}\b|\btitan\s*rtx\b", RegexOptions.IgnoreCase), NvidiaGen.Turing),
            // pascal -- gtx 10-series + titan xp; pascal quadro variants match via the p prefix.
            (new Regex(@"\bgtx\s*10\d{2}\b|\btitan\s*xp\b|\bquadro\s*p\d", RegexOptions.IgnoreCase), NvidiaGen.Pascal),
            // maxwell 2nd gen -- gtx 9-series (gm204/206). gtx 750/750ti was maxwell 1 which didnt have hevc, so we deliberately dont claim it.
            (new Regex(@"\bgtx\s*9\d{2}\b|\btitan\s*x\b", RegexOptions.IgnoreCase), NvidiaGen.Maxwell2),
        };

        private static NvidiaGen DetectNvidiaGen(string name)
        {
            foreach (var (pattern, gen) in NvidiaPatterns)
            {
                if (pattern.IsMatch(name)) return gen;
            }
            return NvidiaGen.PreMaxwell2;
        }

        private static readonly (Regex Pattern, AmdGen Gen)[] AmdPatterns =
        {
            // rdna 4 -- rx 9000-series (forward-looking).
            (new Regex(@"\brx\s*9\d{3}\b", RegexOptions.IgnoreCase), AmdGen.Rdna4),
            // rdna 3 -- rx 7000-series.
            (new Regex(@"\brx\s*7\d{3}\b", RegexOptions.IgnoreCase), AmdGen.Rdna3),
            // rdna 2 -- rx 6000-series + rdna-2 integrated (rembrandt, ryzen 6000u/h).
            (new Regex(@"\brx\s*6\d{3}\b|\bradeon\s+graphics\s*\(rembrandt\)", RegexOptions.IgnoreCase), AmdGen.Rdna2),
            // rdna 1 -- rx 5000-series.
            (new Regex(@"\brx\s*5\d{3}\b", RegexOptions.IgnoreCase), AmdGen.Rdna1),
            // vega / radeon vii.
            (new Regex(@"\brx\s+vega\b|\bvega\s*\d|\bradeon\s+vii\b", RegexOptions.IgnoreCase), AmdGen.Vega),
            // polaris -- rx 4xx / 5xx (570/580/590 etc.). 5xx range conflicts with rdna1s 5000-series, so we match 3-digit 5xx only (not 4-digit 5xxx).
            (new Regex(@"\brx\s*4\d{2}\b|\brx\s*5\d{2}(?!\d)", RegexOptions.IgnoreCase), AmdGen.Polaris),
        };

        private static AmdGen DetectAmdGen(string name)
        {
            foreach (var (pattern, gen) in AmdPatterns)
            {
                if (pattern.IsMatch(name)) return gen;
            }
            return AmdGen.PrePolaris;
        }

        private static readonly (Regex Pattern, IntelGen Gen)[] IntelPatterns =
        {
            // arc discrete + xe-hpg. arc a310/a380/a580/a750/a770 all share the alchemist hevc encode block.
            (new Regex(@"\barc\s*a?\d{3}\b|\bxe-hpg\b", RegexOptions.IgnoreCase), IntelGen.Arc),
            // alder/raptor/meteor lake iris xe in 12th-14th gen.
            (new Regex(@"\b(12|13|14)\d{2}[a-z]?\b.*(uhd|iris)", RegexOptions.IgnoreCase), IntelGen.AlderLake),
            // tiger lake (11th gen) -- xe-lp.
            (new Regex(@"\b11\d{2}[a-z]?\b.*(iris|xe)|\bxe\s*graphics\b", RegexOptions.IgnoreCase), IntelGen.TigerLake),
            // ice lake (10th gen) -- first xe igpu.
            (new Regex(@"\b10\d{2}[a-z]?\b.*iris", RegexOptions.IgnoreCase), IntelGen.IceLake),
            // kaby lake (7th gen) hd 6xx. coffee lake (8th/9th) + comet lake (10th non-ice-lake) reuse the same igpu silicon, so they share this rule.
            (new Regex(@"\b(uhd|hd)\s*graphics\s*6\d{2}\b", RegexOptions.IgnoreCase), IntelGen.KabyLake),
            // skylake (6th gen) hd 5xx -- first with qsv hevc.
            (new Regex(@"\bhd\s*graphics\s*5\d{2}\b", RegexOptions.IgnoreCase), IntelGen.Skylake),
        };

        private static IntelGen DetectIntelGen(string name)
        {
            foreach (var (pattern, gen) in IntelPatterns)
            {
                if (pattern.IsMatch(name)) return gen;
            }
            return IntelGen.PreSkylake;
        }

        private static string SnakeCase(string enumName)
        {
            // PreMaxwell2 -> pre_maxwell2, Rdna1 -> rdna1, KabyLake -> kaby_lake
            var sb = new System.Text.StringBuilder();
            for (int i = 0; i < enumName.Length; i++)
            {
                char c = enumName[i];
                if (char.IsUpper(c) && i > 0 && !char.IsUpper(enumName[i - 1])) sb.Append('_');
                sb.Append(char.ToLowerInvariant(c));
            }
            return sb.ToString();
        }

        // map a win32_videocontroller row to a GpuInfo, or null if it looks like a virtual/basic/fallback adapter we should ignore.
        private static GpuInfo Classify(JObject raw)
        {
            string name = (raw.Value<string>("Name") ?? "").Trim();
            if (name.Length == 0) return null;
            foreach (var skip in IgnoreNamePatterns)
            {
                if (skip.IsMatch(name)) return null;
            }

            Vendor vendor = DetectVendor(name);
            string gen = vendor == Vendor.Nvidia ? SnakeCase(DetectNvidiaGen(name).ToString())
                : vendor == Vendor.Amd ? SnakeCase(DetectAmdGen(name).ToString())
                : vendor == Vendor.Intel ? SnakeCase(DetectIntelGen(name).ToString())
                : null;

            long vram = raw.Value<long?>("AdapterRAM") ?? 0;
            // heuristic: igpus typically expose <2 gb adapterram (rest is shared/dynamic). discrete cards report full vram. 2 gb threshold catches apus/igpus that reserve up to 2 gb dedicated (ryzen) without losing genuinely small discrete cards from ~2014.
            bool isDiscrete = vram >= 2L * 1024 * 1024 * 1024 && vendor != Vendor.Intel;

            return new GpuInfo
            {
                Name = name,
                Vendor = vendor,
                Generation = gen,
                IsDiscrete = isDiscrete,
                VramBytes = vram,
                Driver = (raw.Value<string>("DriverVersion") ?? "").Trim(),
            };
        }

        // every adapter we can usefully classify, discrete cards first.
        public static List<GpuInfo> ListGpus()
        {
            var rows = QueryVideoControllers();
            var result = rows.Select(Classify).Where(g => g != null).ToList();
            var vendorRank = new Dictionary<Vendor, int> { [Vendor.Nvidia] = 0, [Vendor.Amd] = 1, [Vendor.Intel] = 2, [Vendor.Unknown] = 3 };
            result = result.OrderBy(g => g.IsDiscrete ? 0 : 1).ThenBy(g => vendorRank[g.Vendor]).ToList();
            return result;
        }

        // best guess at which adapter obs will use for hardware encoding. null means callers should fall back to software encode.
        public static GpuInfo PrimaryGpu()
        {
            var gpus = ListGpus();
            return gpus.Count > 0 ? gpus[0] : null;
        }
    }
}
